project_badges.go

```diff
@@ -0,0 +1,205 @@
+package gitlab
+
+import (
+	"fmt"
+	"net/url"
+)
+
+// ProjectBadgesService handles communication with the project badges
+// related methods of the GitLab API.
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/project_badges.html
+type ProjectBadgesService struct {
+	client *Client
+}
+
+// ListProjectBadgesOptions represents the available ListProjectBadges()
+// options.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/project_badges.html#list-all-badges-of-a-project
+type ListProjectBadgesOptions ListOptions
+
+// ListProjectBadges gets a list of a project's badges and its group badges.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/project_badges.html#list-all-badges-of-a-project
+func (s *ProjectBadgesService) ListProjectBadges(pid interface{}, opt *ListProjectBadgesOptions, options ...OptionFunc) ([]*ProjectBadge, *Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+	u := fmt.Sprintf("projects/%s/badges", url.QueryEscape(project))
+
+	req, err := s.client.NewRequest("GET", u, opt, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var pm []*ProjectBadge
+	resp, err := s.client.Do(req, &pm)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return pm, resp, err
+}
+
+// GetProjectBadge gets a project badge.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/project_badges.html#get-a-badge-of-a-project
+func (s *ProjectBadgesService) GetProjectBadge(pid interface{}, badgeID int, options ...OptionFunc) (*ProjectBadge, *Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+	u := fmt.Sprintf("projects/%s/badges/%d", url.QueryEscape(project), badgeID)
+
+	req, err := s.client.NewRequest("GET", u, nil, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pm := new(ProjectBadge)
+	resp, err := s.client.Do(req, pm)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return pm, resp, err
+}
+
+// AddProjectBadgeOptions represents the available AddProjectBadge() options.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/project_badges.html#add-a-badge-to-a-project
+type AddProjectBadgeOptions struct {
+	LinkURL  string `json:"link_url,omitempty"`
+	ImageURL string `json:"image_url,omitempty"`
```

Comment:
This fields are requied by gitlab API, why you think that them need to be pointer, can you explain that for me, please?) For what need to add url tag? This fields parse to json only. For future support GET request?)

Thanks for your mentoring.

Reply:
Yeah, this is mostly to keep things consistent. If we would only make optional fields pointers, then it would be harder to use this package. In that case you would have to either know if a field is required or look up the expected value (`*string` or `string`) in the code. So by always using pointers in the option structs users of this packages can have a uniform way to use those structs.

For the `url` tags its again about consistency to prevent errors. If we just always add them on our options structs we don't have to specifically check how that specific struct is used as it will always work.